gcloud-java-datastore/src/main/java/com/google/gcloud/datastore/testing/LocalGcdHelper.java

```diff
@@ -179,49 +179,134 @@ private static Path executablePath(String cmd) {
+        "com.google.apphosting.client.serviceapp.BaseApiServlet";
+
+    private final BufferedReader errorReader;
+    private String currentLog = null;
+    private Level currentLogLevel = null;
+
+    ProcessErrorStreamReader(InputStream errorStream, String blockUntil) throws IOException {
+      super("Local GCD ErrorStream reader");
+      setDaemon(true);
+      errorReader = new BufferedReader(new InputStreamReader(errorStream));
       if (!Strings.isNullOrEmpty(blockUntil)) {
         String line;
         do {
-          line = reader.readLine();
+          line = errorReader.readLine();
         } while (line != null && !line.contains(blockUntil));
       }
     }
 
-    void terminate() throws InterruptedException, IOException {
-      process.destroy();
-      process.waitFor();
-      reader.close();
+    void terminate() throws IOException {
+      writeLog(currentLogLevel, currentLog);
+      errorReader.close();
     }
 
     @Override
     public void run() {
       try {
-        while (reader.readLine() != null) {
-          // consume line
+        boolean errorReaderDone = false;
+        String previousLine = "";
+        String currentLine = "";
+        while (!errorReaderDone) {
+          previousLine = currentLine;
+          currentLine = errorReader.readLine();
+          if (currentLine == null) {
+            errorReaderDone = true;
+          } else {
+            processLogLine(previousLine, currentLine);
+          }
         }
       } catch (IOException e) {
         // ignore
       }
     }
 
-    public static ProcessStreamReader start(Process process, String blockUntil) throws IOException {
-      ProcessStreamReader thread = new ProcessStreamReader(process, blockUntil);
+    private void processLogLine(String previousLine, String currentLine) {
+      // Each gcd log is two lines with the following format:
+      //     [Date] [Time] [GCD_LOGGING_CLASS] [method]
+      //     [LEVEL]: error message
+      // Exceptions and stack traces are included in gcd error stream, separated by a newline
+      Level nextLogLevel = getLevel(currentLine);
+      if (previousLine.contains(GCD_LOGGING_CLASS) && nextLogLevel != null) {
+        writeLog(currentLogLevel, currentLog);
+        if (currentLine.startsWith("SEVERE: ")) {
+          // don't show duplicate error messages from gcd.sh (see issue #258)
+          currentLog = null;
+          currentLogLevel = null;
+        } else {
+          currentLog = "GCD" + currentLine.split(":", 2)[1] + System.getProperty("line.separator");
+          currentLogLevel = nextLogLevel;
+        }
+      } else if (currentLog != null && currentLog.length() > LOG_LENGTH_LIMIT) {
+        // log processing may be off, so drop some logs before the string becomes too big
+        currentLog = null;
+        currentLogLevel = null;
+      } else if (currentLog != null && isUsefulLogInfo(currentLine)) {
+        currentLog += currentLine + System.getProperty("line.separator");
```

Comment:
I think if currentLog != null (which should mean a log message written by GCD_LOGGING_CLASS) we always want to display the message, no?

basically each time you detect a level, if prior string has the class we are in a "collecting mode" until we detect another level (though still in "collecting" mode we drop lines if 
we exceed size X). When we detect another level we display prior and again reset the "collecting mode" based on previous line. 

Also, adding the lines should be done using "previousLine" and only after we know we didn't see a new level.